### ISSUE TITLE
Extending "PETE" with Reusable Components So We Can Build People-Centric Standards-Compliant Data-Driven Apps Easier/Faster/Reliably

### ISSUE THREAD

**Issue by nelsonic** [dwyl/technology-stack#67]
# Context 🤔 

We have been using the ["**PETE**" Stack](https://github.com/dwyl/technology-stack/issues/37) for the past **18 Months** and it has been a really _good_ experience 
for end-users (of the apps we've built), developers, product owners/clients. 
We _love_ Phoenix and Elixir and feel it's been a _good_ choice of framework and programming language.

My/our _one_ "***regret***" is not _pushing_ for more **modularity** earlier on 
which has meant  that we have "locked up" _most_ of the useful functionality in the Apps we have built
and thus have to "re-implement" lots of "boilerplate" each time we start a new project. 😞 

## _Example_ 💊 

A _good_ example of the PETE stack is Healthlocker https://github.com/healthlocker/healthlocker 
The end-users like using and is _much_ better than anything the NHS had before (_for this use-case_). 🥇 
Everyone who worked on the project can be _proud_ of the work that has been done. 🎉 

_However_ if you look at the [**`mix.exs`**](https://github.com/healthlocker/healthlocker/blob/becff6feafb3e10ebe061f7056bcc04bf4755c0c/mix.exs#L55) file, you will see that while we used several open source modules created by other people, we did not _create_ any reusable components/packages/modules ourselves ...
That's "OK" because it was not the (_stated_) "goal" of the project to "_create **reusable code**_", 
the goal of the project was to create an App that helps people "self-manage their [mental] health" and "effectively communicate with their clinicians". 

_However_ it was an aim of the project to be ***Open Source*** so that _anyone_ could contribute!
And while we "checked the box" of Open Source by making the project **`public`** on GitHub,
we have not had any _contributions_ from anyone _outside_ of the "core" team. 
This raises a _separate_ question around what constitutes "success" in Open Source: https://github.com/dwyl/technology-stack/issues/66 ...

Our _hypothesis_ is that if we had created Healthlocker as a collection of reusable components which then get _assembled_ into an App, we _could_ have aided the "core" goal of the project because by making reusable components, each time one of those components is reused and _improved_, the _original_ project receives all the improvements "for free". 

By creating Healthlocker as a one "monolithic" repository (_AKA "**monorepo**"_) 
we were able to move fast(er) _initially_ and deliver the App to users/stakeholders quicker ... 🐎 
but _now_ when we need to implement Auth, Permissions or Chat in a new project 🆕 
we have to do it all "from scratch" ... ⏳ 

## _Analogy_: Smart Phones 📱 

Think about the Smart/Mobile Phone in your hand/pocket, it's made up of [_millions_ of components](https://fossbytes.com/whats-inside-smartphone-depth-look-parts-powering-everyday-gadget/).
The software stack that goes into making the smartphone possible is _billions_ of lines of code.
Each time one of the elements in the software stack (Kernel, APIs, UI Libraries, Build Tools) improves, 
_everyone_ in the ecosystem (developers, users, companies) benefits from the improvement.
If the stack of software for the smartphone was a "monorepo" it would have failed _long_ ago;
because it would have become unmaintainable and no _individual_ person would be able to understand it!

> **Note**: For this analogy, focus on the _software_ and ignore the fact that for the most part the individual hardware components (screen, battery, processor, camera etc.) are not upgradeable. 🙄 

## Lessons Learned 📚 

We have _all_ learned a lot from our work over the last few years. A few lessons
+ [ ] The "flexibility" of "NoSQL" rapidly becomes a burden when data is inconsistent and queries slow!
  > We felt this pain when we tried to use ElasticSearch as a _primary_ DB, Redis to store user data, or JSON in PostgreSQL which resulted in painfully slow queries! MongoDB was a *pig* when we used Meteor; they sacrificed reliability for "bigness".
+ [ ] Don't delay/procrastinate on figuring out the schema for your App. 
The sooner you think in "data structures" the better. see: https://github.com/dwyl/learn-elm/issues/120
+ [ ] **_Embrace_ Email** as a means of connecting/communicating with and getting _feedback_ from users!
+ [ ] **_Proactively_ Implement Analytics** and capture metrics for every aspect of your app(s)! 
  > Trying to _retrospectively_ apply "GTM" as an _afterthought_ is like using a tennis racket to paddle a canoe _upstream_! It might "kind work" but you'll ultimately "lose" drowning in data noise!

# Proposal 💡 

Our mission for the rest of 2018 is to create the reusable stack we will be deploying in 2019:

![dwyl-technology-stack-2019](https://user-images.githubusercontent.com/194400/47501921-808c3d80-d85e-11e8-8db0-3a779721e421.png)

> Diagram: https://drive.google.com/file/d/1gfvgK1Mn7EuamHJroS8G7FUqauRgA9EH
Note: the colors of the chevrons are not significant. they are merely to help differentiate between the different elements in the stack. If anyone wants to make this look better, please go for it!

If this _feels_ like "a lot", don't be discouraged or overwhelmed, we already have a _massive_ head start!
A _lot_ of the work on this has already been started and in some cases shipped!
Break it down into "chunks" and start from the bottom: the Append-only Log is the _core_ of all apps.
see: [https://github.com/dwyl/**phoenix-ecto-append-only-log-example**](https://github.com/dwyl/phoenix-ecto-append-only-log-example) (_if you haven't already_)

@Danwhy is already making great progress with Append-only Log: https://github.com/dwyl/alog
Both @Cleop and @RobStallion have made great contributions to https://github.com/dwyl/autoform
and we are already _actively_ ["dogfooding"](https://en.wikipedia.org/wiki/Eating_your_own_dog_food) elements in our client projects!

We'll build a "minimalist" analytics system "from scratch" on top of alog: https://github.com/dwyl/atm
We have done _some_ work towards this in: https://github.com/dwyl/hits 

Auth will need to be assembled from the ground up based on alog, fields and autoform. 
So for now, _ignore_ the work that had been previously done. 

Once we have the "base" layer of alog, Fields, ATM, Autoform, Auth and Admin, we can build the more "feature rich" tools: Contact and Feedback. 

Hopefully this will clarify the "strategic" direction we are taking when building our most recent projects in a _modular_ way. 

# Name? 💭 

For now, we don't have a _name_ for the _collection_ of modules/packages.

>  Open to suggestions as to what we should call this collection of modules. 

I'd love to call this the "**_Obvious_ Stack**" because most of these things are _obviously_ useful to most apps we have (already) built and will give us a _massive_ head start for anything we build in future. 

Considering "_repurposing_" https://github.com/dwyl/abase as it was the "spiritual grandfather" of this.
"**aBase**" was built and used used for a couple of Node.js Apps in 2016/17 but not developed further, as transitioned to Phoenix in early 2017. 
It's not the _best_ name (mea culpa), but at least it conveys the meaning of what we are building.

# Feedback / Thoughts / Questions / Discussion ? 💬 

As always _your_ feedback is very much invited/requested/appreciated!
If you have ideas, thoughts, questions, concerns, please share!!

Thanks! ✨ 

***Related*** to: https://github.com/dwyl/learn-elm/issues/121 (Creating Elm Packages) ❤️ and https://github.com/dwyl/hq/issues/497 (Learning Goals). 🎯 
***Required*** by: https://github.com/dwyl/feedback/issues/96 (Feedback Widget) and https://github.com/dwyl/product-roadmap/issues/7 (Metrics)

**Comment by iteles** [comment]
> Our hypothesis is that if we had created Healthlocker as a collection of reusable components which then get assembled into an App

This. This is a huge lesson from the work we have done over the past 2 years. Our small modules (which we _barely_ use ourselves) attract a fair amount of attention and contributions - the highly modular approach makes a lot of sense.

Having now done this on our latest two client projects, this way of working makes my heart sing ❤️ 🎶 It's better for us, better for our clients and better for the open source community.

Thanks @nelsonic for the proposal diagram too, having it so clearly laid out is extremely useful!

**Comment by Cleop** [comment]
Thanks for writing this up @nelsonic, I'm enjoying contributing to its progress and looking forward to seeing it take shape!

**Comment by RobStallion** [comment]
@nelsonic This is awesome. I agree wholeheartedly. ❤️ ❤️ 

> Our hypothesis is that if we had created Healthlocker as a collection of reusable components which then get assembled into an App, we could have aided the "core" goal of the project because by making reusable components, each time one of those components is reused and improved, the original project receives all the improvements "for free".

This point especially rings true for me. Although I did a lot of work with healthlocker, when I have needed to come back to it to make improvements/fix bugs/add a new feature, I find it time consuming and honestly a little daunting. The great thing about the modular approach and components is that when we need to upgrade a component, everywhere that component is used is also updated. 

Having to manually find and update everywhere in an application that you have done something specific can be time consuming and error prone. 

I really enjoy the modular approach. There a plenty of reasons for this, the top few being that it saves time on a project if we can reuse something other developers have already built (and we can use that time on the projects rather than reinventing the wheel) and code is in smaller, more manageable chunks (and often more robust as many people have thought long and hard about the best way to implement that one thing). 

My favourite reason though is that I feel like the code is much more accessible. What I mean by this is that when it comes to big monolithic projects, they could have (and often do have) plenty of great functions and modules that have been built into them but it is hard to know what they will have and where to look for it. With the modular approach I feel it is much easier to find exactly what it is that you are looking for, apply it to your use case and (maybe most importantly) make suggestions for how the code could be improved/enhanced.

**Comment by SimonLab** [comment]
Agree :+1:  if we are able to split applications in multiple reusable components, the creation process will be much faster and enjoyable. For example rewriting the authentication logic for each new application is time consuming and tiring!

However I'm wondering if any features can be implemented with reusable components or do we need to have a limit on the kind of projects that can be implemented with modules? 

There is a risk to try to adapt the reusable code to match some specific features of a project. What would be an acceptable solution, accept any PRs in the reusable modules? This might pollute the API and the modules might not be as easy to reuse. Fork the modules to make them specific enough for the application? This will break the idea of reusable component.
I had this issues on previous project where small PRs where added to some npm packages to just be able to add/update some features on the project. So to avoid this we want to make sure from the start that all the features can be implemented easily with the components or we need to direct and think of new features based on what the components can do instead of letting the product owner decide/impose what they have in mind.

It's difficult for me to have a precise idea on how everything will work together, but I like this macro vision on how to build our next projects. A simple example which demonstrate how to use the "PETE extended" would help us to see what are the limitations but also confirm our motivations to go in this direction.

**Comment by nelsonic** [comment]
@SimonLab good question(s)! 
Provided we have a definition of all types of data in **`fields`** we can build _anything_!
(_I don't think we need to impose a "limit" yet. Unless you can think of an example ..._)

> We can discuss _specific_ field types in issues in **`fields`**, for _example_: **`postcode`**.
What should the _validation_ for the **`postcode`** field be? 
Should we have a RegEx that matches UK postcodes or USA?
Should **`fields`** receive the browser's Language String e.g: `EN/US` in order to determine which RegEx to use? ***or*** should we be able to pass a **_custom_ validation** to **`fields`** to make it more flexible? We can discuss this on a case-by-case basis as the need arises; we don't need to ["waterfall"](https://en.wikipedia.org/wiki/Waterfall_model) the whole thing thing _up-front_.

The only risk I see is in defaulting to the "old habits" of building _all_ the functionality into the "Client App" instead of being _disciplined_ and putting the functionality in a _reusable_ package.

We can only _build_ a simple example once the `alog` and `fields` are in a _useable_ state.
The "simple" example is going to be "ATM" which will be a _useable_ Analytics package we will use to make "Hits" see: https://github.com/dwyl/product-roadmap/issues/7 a "real world" App with _real_ users. From there we can iterate "up the stack".

Yes, we've done things a little bit "in reverse" with our latest client project because **`fields`** is not (_yet_) built so we are not using it (_yet_), but that will change v. soon.

Our _first/next_ step is to make the docs for https://github.com/dwyl/alog _super_ beginner friendly.
We have a _few_ good examples of this in @dwyl e.g: https://github.com/dwyl/hapi-auth-jwt2
But we should not _limit_ our "inspiration" for "good docs" to @dwyl ... we should find the _best_ docs for _any_ Open Source project and _match/beat_ it!

> Functionality is _essential_, but docs and examples 
are the difference between "that's nice" and "**this is _amazing_**!

Don't get "hung up" on the details of how the more advanced features like Admin and Contact will work or look. focus on the "core" components at the "base" of the stack for now.

If it's "difficult" to have an idea, simply go look at 
+ https://rubyonrails.org 
+ https://laravel.com 
+ https://www.djangoproject.com

We are re-imagining the Web Application Stack from the ground-up to be distributed, fault-tolerant, real-time, offline-first, metrics-driven, GDPR-compliant and _fun_ to use for developers and end-users.
(_if that sounds like a lot of "buzzwords", just focus on "Hits" as the example for now..._) 😉

**Comment by Danwhy** [comment]
Thanks for this write up @nelsonic, it's really useful to see the full plan of what we're working towards. I've really been enjoying working on the various parts of this so far, and am looking forward to those we haven't got to yet.

The only question I have is around how all of these modules actually fit together. Based on the above, it looks like each element in the stack depends on those below it, but my understanding was that each of these modules would be completely independent of each other; so if somebody wanted to use only the `analytics` module, or only the `auth` module, they wouldn't be required to use the `append only log` if that wasn't what they needed.

You also mentioned repurposing `abase`, so would that be the "link" between all of these modules? (So in our client apps we'd essentially just require `abase`, which would give us the full functionality of the stack, in a framework that's essentially "Phoenix+", but the modules would also be available individually for other people that didn't want our whole stack)

I just want to be sure I know exactly how independent these modules are going to be, to make sure I'm implementing them correctly as we go on.

**Comment by nelsonic** [comment]
@Danwhy thanks for the insightful feedback and clarifying questions. 🥇 

_Great news_ that you've been enjoying diving into building the **`alog`** and **`autoform`** pieces. 🎉 

# _Answers_

Each "core" piece should be _independently_  tested and documented. 
I consider **`alog`** and **`fields`** to be the most _decoupled_ "core" pieces
and if someone wants to use **`alog`** and/or **`fields`** _independently_ they are welcome/encouraged to!

The further _up_ the stack the more "coupled" the pieces become to the "core". 
But that does _not_ mean we should build _everything_ into a monolith called "starter-kit" that contains _everything_... because it would _rapidly_ become "bloated"! Instead what we want to do is build each piece as a "lightweight" MVP of the desired functionality so that we can have people in the community contributing to different elements of the stack without necessarily needing to _understand_ the whole thing. (_although our mission is to make the Docs "so good a **complete beginner** can understand_!)

We are _not_ going to build **`auth`** to be "_general purpose_" for _any_ Phoenix project; it will have a "hard" dependency on **`autoform`** which in turn will be _dependent_ on **`fields`**. And since our chosen "data model" for _all_ future projects is an "append-only log" (_for accountability, analytics and distributed systems_), what we are currently calling "**`alog`**" (_see: https://github.com/dwyl/alog/issues/2 for naming discussion_) will _also_ be a "hard" (_not optional_) dependency. 

> Note: Arguably, Auth _should_ be a _completely_ separate ["Umbrella App"](https://elixir-lang.org/getting-started/mix-otp/dependencies-and-umbrella-projects.html) that is deployed _independently_ of any "Client App" logic and just returns a "Session Token" to represent the _session_ for a given user when they are logged in. But deploying multiple apps is _needless_ complexity at this point, and we can always refactor into an Umbrella App later on as the need arises. My only _reasoning_ for doing it _sooner_ would be to have _complete_ separation between Application Data and Personally Identifiable Info such that a "compromise" of the _Anonymised_ Application Data will not give the attacker access to people's PII ... 🤔 

In our "Client" apps there will be no such thing as a _locally_ defined ["User Schema"](https://github.com/club-soda/club-soda-guide/blob/7cf212ee4d1e9a8c3c433e7825f4d46741eb258f/lib/cs_guide/accounts/user.ex#L6-L8).
The concept of a **`person`** will be pre-defined in **`auth`** and if the Client App needs to add a _custom field_ to the **`person`** schema it will be done using a **`person_custom_fields`** "lookup table". e.g: 

| `inserted_at ` | **`attribute_id`** (PK) | `person_id (FK)` | `field_type` | `data` | 
| ----------- | ----------- | ----------- | ----------- |  ----------- |
| 1541609554 | e096d10048146852 | e2ad11e8871e | favourite_color | rainbow | 
| 1541609876 | ab4362a379819971 | 31e711e89bd8 | favourite_food | ice cream |

This will make the framework _incredibly_ flexible and essentially allows people to "drag-and-drop" from a list of "available fields" into a "custom fields" table, thus the "_Product Owner_" can construct their "App" based on reusable components (_think lego bricks!_) without having to write a _line_ of code. "Custom Autoforms" will do the "heavy lifting" to render the forms, validate and store the data and **`admin`** will help the Product Owner view the data as it is submitted.
_Yes_, this will effectively be a way to create a TypeForm/SurveyMonkey/WuFoo with just a few clicks! 😮 

> Note: in this example, we are displaying the data in `plaintext` but it would still be stored encrypted.
> Also, we will _eventually_ have a field definition in **`fields`** for the most _obscure_ fields, but a **`Fields.color`** should be an `encrypted_INT` of [RGBA](https://stackoverflow.com/questions/785551/how-could-i-store-a-color-in-a-database-field) and the **`autoform`** for _inputting_ the data should be a "Colour Picker" widget ... see how we can _rapidly_ go down a "rabbit hole" of thinking about how to implement "all the things" ...? We need to be _disciplined/focussed_ to _only_ build the **`fields`** we _need_ as we need them. 🙄 

It's going to be _much_ easier to explain the reasoning for each element in the "stack" once the _need_ arises. But here's **a basic _example_**: (_"abase-ic", see what we did there...?_ 😉)

# _Example_
 
Imagine you are building a Social Network for Pet Lovers and you need to store specific data for each member of the platform including their pet's **name**, **species**, **breed** and **date of birth**.

The "base" **`person`** schema will be the "_standard_" one defined in **`auth`**:

| `inserted_at ` | **`email_hash`** (PK) | `email_encrypted` | `name_encrypted` | `password_hash` | verified |
| ----------- | ----------- | ----------- | ----------- |  ----------- |  ----------- | 
| 1541609554 | 48146852e096d100 | e2ad11e8871e | b4f54f713284 | 4362a31e711e | 1541609554 |
| 1541609876 | 379819971ab4362a | 31e711e89bd8 | 9566834fc199 | 1ab4362a3456 | null |

> **_All_ personal data** is **_encrypted_** "pre-DB-insertion", so this illustrates data "at rest",
> _also_ hashed/encrypted data will be _considerably_ longer, I've truncated it here for brevity.
> If anyone **`else`** is following along scratching their head as Why or How this is done, 
see: https://github.com/dwyl/phoenix-ecto-encryption-example
> ***Yes***, encrypting all personally-identifiable data is _tedious_ and gives us the _developer's_ "more work", but the _whole_ point of **`fields`** is that the encryption/decryption happens _automatically/transparently_ in the **`load`**  (Ecto Custom Type callback) function and the data is **`plaintext`** within the App and UI. This is the _responsible_ way of building Apps, and "in the future" ***all*** Apps will be built this way _by **`default`**_; _security & privacy first_!

The **`pets`** table would be:

| `inserted_at ` | **`person_id`** | `pet_id` | `name_encrypted` | `species` | `breed` | `date_of_birth_encrypted` |
| ----------- | ----------- | ----------- | ----------- |  ----------- |  ----------- |  ----------- | 
| 1541609554 | 48146852e096d100 | 8871ee2ad11e | 13284b4f54f7 | 4362a31e711e | 123 | 76834fc199 |
| 1541609876 | 379819971ab4362a | 89bd831e711e | 4fc199956683 | 1ab4362a3456 | 345 | 4362a3456 |

These fields might be _rendered_ on the _person_'s "profile page" as:
+ Name: Alex
+ Age: 27
+ Pet: Fluffy Bandit
  + Species: [Dog](https://en.wikipedia.org/wiki/Dog)
  + Breed: [Cockapoo](https://en.wikipedia.org/wiki/Cockapoo) 
  + Age: 3 years ([28 Dog Years](https://pets.webmd.com/dogs/how-to-calculate-your-dogs-age))

> Ignore the _other/advanced_ features of a "Social Network" for now and just focus on the fact that the **`person`** has this _related content_ in the **`pets`** table. The person can have as many pets as they like; **one-to-many**. But we are not _extending_ the **`person`** record, we are creating a _separate_ table with the custom fields.

A _naive_ way of implementing this "social network for petlovers" would be to add custom fields to the person ... e.g:
+ pet1_name
+ pet1_species
+ pet1_breed
+ pet1_dob
+ pet2_name
+ pet2_species 
+ etc ...

That would be a _non-relational_/[denormalised](https://en.wikipedia.org/wiki/Database_normalization) way and would mean that the "_outlier_" on the network who _literally_ owns a [zoo](https://en.wikipedia.org/wiki/Dartmoor_Zoological_Park) cannot list all their pets ...

As you can tell from the data we are capturing for the **`pet`**, we can _easily_ construct the **`pets**` table using **`fields`**:
+ `name_encrypted`: `Fields.name_encrypted`
+ `species`:
+ `breed`: `Fields.
+ `date_of_birth_encrypted`: `Fields.timestamp_encrypted`

> **Note**: we could further ["**normalize**"](https://en.wikipedia.org/wiki/Database_normalization) the **`pets`** data by creating an **`animals`** lookup_table which would store the species, breed, sub-type variant data ... see: https://en.wikipedia.org/wiki/Phylogenetic_nomenclature thus the _reference_ in the **`pets`** table would be to a PK in the **`animals`** lookup_table. 
But, again for illustrative purposes we are keeping this "simple".

# Repurposing "Abase"?

Yes, having a "Phoenix+" stack (_name TBD_) which would include "users" (**`people`**), `emails_sent`, **`auth`**, **`admin`**, **`analytics`**, etc. is the _medium-term_ goal for this _quest_. 

Being able to reference _one_ package in the "Client App" `mix.exs` dependencies would be _slick_.

We need _help_ on defining the _name_ for this Application Framework ["_Scaffolding_"](https://www.powerthesaurus.org/scaffolding) ... `#HelpWanted`
Sadly, as usual in computer science / software, most of the "good" names are "taken" ...

The only _reason_ I like **`abase`** is because it describes what it is/does it's the "base" for building other apps, and _alphabetically_ it would appear _first_ (_or near the top_) of any "list" of frameworks.
But I don't think we are "married" to the name _at all_! If anyone can think of a _better_ name, please do!

<hr />

# tl;dr

> **Please note**: this is not a "_full plan_", it's merely an _initial_ diagram to help the team visualise/understand that there _is_ a "vision" for where we are going 
and a "logical order" for building things. 
The "_destination_" is something akin to a "create-app" tool that includes _everything_ for a person to get started with building their App. 
We are nowhere _near_ "done" with **`alog`** for example, by the time it's at **`1.0`** it needs to be the _de facto_ way of building _distributed/offline-first_ apps in Phoenix and we need to _show_ people _Why_ using a "traditional" CRUD is a _bad idea_ for apps that allow _off-line_ record creation and want to have accountability built-in.

> Imagine you're building a Todo List with Time Tracking and you want people to be able to add items to their list while they are on the London Underground (without network access), Having and auto-incrementing integer as the Primary Key for the "Task" Schema will mean that the person _cannot_ create records _offline_ and then sync them back up to the DB when they get back online.
This means we're going to need to create something _similar_ to [PouchDB](https://pouchdb.com) for use in our Progressive Web Apps so that people can _store_ data on the device, search for history and create **`new`** records on their device. (It means we'll need to _simulate_ the Ecto Query Syntax on the Client!) And then _sync_ should be _automatic/transparent_ once network access is re-gained. 
This is an _obvious_ "gap" in the diagram (_above_) but it's _100% deliberate_ because if we were to map out the _whole_ diagram it would look more like a "mind map" and would likely "_overwhelm_" people. One of the most important things in "application architecture" is to avoid over-complicating things up-front. 

> We are working towards a future of ["flow-based programming"](https://youtu.be/up2yhNTsaDs?t=229) where we have a _bunch_ of "lego bricks" that can be combined visually into building _anything_. It will allow _complete_ beginners to drag-and-drop **`fields`** to create questionnaires to capture data in _minutes_.

> By the time we are "done" we will be able to _visualise_ data flowing through the system in real-time.
Product owners will be able to create [A/B & multivariate tests](https://github.com/dwyl/learn-ab-and-multivariate-testing) and track the data/user-feedback simply by re-arranging the **`fields`** in a "custom" **`autoform`** layout. Running 50:50 or 10:90 traffic through the "experiment" version will be **3 clicks**. 

> My _personal_ goal is to use this stack for capturing "farming data" in https://github.com/dwyl/home and writing ML algorithms to "understand" how plants grow and how to _automate_ nutrient dispensing for optimal growth/health.

> We just have to _resist_ the temptation to over-think things too early and avoid spending too much time dreaming about ["the future"](https://www.youtube.com/watch?v=8pTEmbeENF4) ... we must _focus_ on building _small_ Example/Demo apps with the stack and _evolving_ the stack as needed. More on that v. soon! 🔜 

# _Real_ App Examples for Learning the Stack?

My **_personal_ learning objective** (see: https://github.com/dwyl/hq/issues/497) in the next 6 months 
is to create one _fully functional_ end-to-end demo/example app using the "New Stack" each month. 

Culminating in a "_curriculum_" of "**7 Apps in 7 Weeks**":  (_from my personal notebook/journal..._)
![seven-apps-in-seven-weeks](https://user-images.githubusercontent.com/194400/48224700-86bc0700-e392-11e8-8ea4-43e08828e9bc.JPG)

that _anyone_ can use to learn how to build using the "New Stack" from _anywhere_!
Then we "_just_" need to create the learning content for: 
+ 1 week of **HTML** including Forms, Browser Validation and "understanding data", 
+ 1 Week of **CSS/Tachyons** > how to build a Medium Blog Clone
+ 1 Week of **Design/Layout/Typography**, 
+ 1 week of **Analytics and Data-driven Decision Making**
+ 1 week of **User Experience Testing**, Feedback collection and Analysis
+ 1 week of **JavaScript** with TDD: https://github.com/dwyl/todomvc-vanilla-javascript-example (_which is **Why** I'm so keen to get **feedback** from everyone ..._)
+ 1 week of "**Data & Database Concepts**" with an "Address Book", "Movie Database" and "Social Network" examples ...

And You can _begin_ to see where this is going ... 😉 
We will be _each_ be building one or more "example apps" as _complete_ learning resources in the coming weeks/months. And in the _course_ of building these examples, we will _evolve_ the "stack".
Then, armed with all the _examples_ we will be able to build our _own_ App(s) and any "Client Apps" ...
see: https://github.com/dwyl/start-here#what 😮 

![carl-sagan-nature-quotes-if-you-wish-to-make-an-apple-pie-from](https://user-images.githubusercontent.com/194400/48225347-6bea9200-e394-11e8-9249-d6d316c8de0a.jpg)

**Comment by nelsonic** [comment]
The next element in the stack is Logging. 🪵 
So that's what I'm working on _now_. 🧑‍💻 
That way I can re-use it in all the other projects ASAP. ⏳

**Comment by nelsonic** [comment]
Related: https://github.com/dwyl/technology-stack/issues/172
